vllm/v1/core/kv_cache_manager.py

```diff
@@ -212,22 +216,40 @@ def allocate_slots(
         if not self.enable_caching:
             return new_blocks
 
-        # NOTE(rickyx): We are assuming the `num_tokens` are actual
-        # tokens rather than lookahead slots (e.g. for speculative decoding).
-        # TODO(rickyx): When supporting speculative decoding, we will need to
-        # differentiate between them so that we can know how many blocks are
-        # full after appending the actual tokens.
-        num_full_blocks = (num_computed_tokens + num_tokens) // self.block_size
-        num_computed_full_blocks = num_computed_tokens // self.block_size
-        new_full_blocks = req_blocks[num_computed_full_blocks:num_full_blocks]
+        # Calculate the total number of complete blocks needed after appending
+        # new valid tokens.
+        max_speculative_tokens = max_speculative_tokens or 0
+
+        # We subtract max_speculative_tokens from num_computed_tokens to
+        # get the least number of tokens in the KV cache.
+        # All tokens before this number are guaranteed to be valid and
+        # stored in the cache.
+        min_num_last_step_computed_tokens = num_computed_tokens \
+                                - max_speculative_tokens
+        # Speculated tokens generated in the last step are counted in
+        # request.num_computed_tokens. Specualted tokens in the current step
+        # are not counted and cached because they are not verified/accepted
+        # yet.
+        num_tokens_wo_spec_tokens = num_tokens - max_speculative_tokens
+
+        min_num_last_step_computed_full_blocks = \
+                        min_num_last_step_computed_tokens // self.block_size
+        num_full_blocks_after_append = (
+            num_computed_tokens + num_tokens_wo_spec_tokens) // self.block_size
+
+        new_full_blocks = req_blocks[min_num_last_step_computed_full_blocks:
+                                     num_full_blocks_after_append]
```

Comment:
Could you split the logic into two cases: with and without spec decoding? This way, readers can skip the complex logic here when they are not interested in spec decoding.